### ISSUE TITLE
[Bug] Dynamic filtering drops accepted terminal rollouts when replacement prompts are exhausted

### ISSUE THREAD

**Issue by ai-yang** [OpenRLHF/OpenRLHF#1310]
## Summary

When dynamic filtering rejects a rollout group after the prompt loader has been exhausted, `SamplesGenerator._generate_vllm()` returns an empty list immediately. This discards every rollout already appended to `accepted_experiences` and cancels all remaining vLLM refs, including refs whose generated rollouts may be valid.

**Suggested severity: medium-high.** Dynamic filtering is opt-in, but it is enabled by shipped DAPO/REINFORCE recipes and the failure silently loses completed generation work. On a small dataset, the only generation batch can become empty and produce zero optimizer updates. Both synchronous and asynchronous trainers use this shared generator.

Tested on `main` at `3c3be6234e0cb353e76bb8019947db9dfe99fca7` (`v0.11.0`).

## Root cause

The generator correctly accumulates accepted results at [`samples_generator.py:179-182`](https://github.com/OpenRLHF/OpenRLHF/blob/3c3be6234e0cb353e76bb8019947db9dfe99fca7/openrlhf/trainer/ppo_utils/samples_generator.py#L179-L182). If a later result is filtered, it asks the prompt loader for a replacement. When no replacement exists, [`samples_generator.py:187-190`](https://github.com/OpenRLHF/OpenRLHF/blob/3c3be6234e0cb353e76bb8019947db9dfe99fca7/openrlhf/trainer/ppo_utils/samples_generator.py#L187-L190) does this:

```python
if exhausted and not new_prompts:
    for remaining_ref in pending_refs:
        ray.cancel(remaining_ref)
    return [], prompts_consumed, True
```

Returning `[]` throws away `accepted_experiences`. Cancelling `pending_refs` can also discard already-generated or soon-to-complete accepted groups. This conflicts with `_collect_prompt_batch()`'s documented contract that callers must process a partial batch collected before exhaustion.

## Reachability through shipped recipes

[`examples/scripts/train_dapo_ray_hybrid_engine.sh`](https://github.com/OpenRLHF/OpenRLHF/blob/3c3be6234e0cb353e76bb8019947db9dfe99fca7/examples/scripts/train_dapo_ray_hybrid_engine.sh#L17-L33) enables dynamic filtering with range `(0, 1)`, `rollout.batch_size=128`, `n_samples_per_prompt=8`, and `data.max_samples=20000`. Its shipped [example reward function](https://github.com/OpenRLHF/OpenRLHF/blob/3c3be6234e0cb353e76bb8019947db9dfe99fca7/examples/python/reward_func.py#L23-L30) generates independent binary scores.

For that example, `20000 % 128 == 32`, so the final initial prompt batch contains 32 groups. A binary group of eight samples is filtered when all scores are zero or all are one, with probability `2 / 2^8 = 1/128`. The probability that the terminal 32-group batch contains at least one filtered group is therefore approximately `22.2%`. When that happens, the current early return empties the whole terminal result instead of returning its accepted groups.

## Deterministic CPU reproduction

The following script loads the real method from the checkout and only stubs Ray/vLLM transport. It deterministically completes one accepted group, then one rejected group, while another accepted group remains pending.

```python
import importlib.util
import sys
import types
from pathlib import Path
from types import SimpleNamespace as NS

import torch


class Pbar:
    def __init__(self, *_args, **_kwargs):
        pass

    def set_postfix(self, *_args, **_kwargs):
        pass

    def update(self, *_args, **_kwargs):
        pass


ray = types.ModuleType("ray")
ray.cancelled = []
ray.wait = lambda refs, **_: (refs[:1], refs[1:])
ray.get = lambda ref: ref
ray.cancel = ray.cancelled.append
sys.modules["ray"] = ray

vllm = types.ModuleType("vllm")
vllm.SamplingParams = object
sys.modules["vllm"] = vllm
sys.modules["tqdm"] = NS(tqdm=Pbar)

experience_module = types.ModuleType("openrlhf.trainer.ppo_utils.experience")
experience_module.Experience = object
sys.modules[experience_module.__name__] = experience_module

engine_module = types.ModuleType("openrlhf.trainer.ray.vllm_engine")
engine_module.batch_vllm_engine_call = lambda *_args, **_kwargs: None
sys.modules[engine_module.__name__] = engine_module

logging_module = types.ModuleType("openrlhf.utils.logging_utils")
logging_module.init_logger = lambda *_args, **_kwargs: NS(info=lambda *_args, **_kwargs: None)
sys.modules[logging_module.__name__] = logging_module

path = Path("openrlhf/trainer/ppo_utils/samples_generator.py")
spec = importlib.util.spec_from_file_location("repro_samples_generator", path)
module = importlib.util.module_from_spec(spec)
sys.modules[spec.name] = module
spec.loader.exec_module(module)


def group(prefix, scores):
    return [NS(name=f"{prefix}-{i}", scores=torch.tensor([score])) for i, score in enumerate(scores)]


refs = [
    group("accepted", [0.0, 1.0]),
    group("rejected", [0.0, 0.0]),
    group("pending", [0.0, 1.0]),
]

generator = module.SamplesGenerator.__new__(module.SamplesGenerator)
generator.args = NS(algo=NS(dynamic_filtering_range=(0.0, 1.0)))
generator._dispatch_prompts_to_vllm = lambda *_args, **_kwargs: refs
generator._process_response_into_experience = lambda response, **_kwargs: response

loader = iter((None, [f"prompt-{i}"], [None], [None]) for i in range(3))
experiences, prompts_consumed, exhausted = generator._generate_vllm(
    loader, num_prompts=3, dynamic_filtering=True
)

print("returned:", [experience.name for experience in experiences])
print("prompts_consumed:", prompts_consumed)
print("exhausted:", exhausted)
print("cancelled:", [[experience.name for experience in ref] for ref in ray.cancelled])
```

Actual output on current `main`:

```text
returned: []
prompts_consumed: 3
exhausted: True
cancelled: [['pending-0', 'pending-1']]
```

The first accepted group has already been added to `accepted_experiences`, but it is missing from the return value. The pending group is valid too, but it is cancelled.

## Expected behavior

If no replacement prompt exists, the generator should finish collecting refs that were already dispatched and return every accepted terminal rollout. For the reproduction above:

```text
returned: ['accepted-0', 'accepted-1', 'pending-0', 'pending-1']
prompts_consumed: 3
exhausted: True
cancelled: []
```

## Suggested fix

Remove the exhaustion early-return block. With no replacement, the existing loop can naturally drain `pending_refs`; the existing final return then preserves `accepted_experiences` and the correct `exhausted=True` state.

Returning the already-accepted list immediately would still discard potentially valid pending refs, so draining the refs is preferable.

## Duplicate check

I searched open and closed issues and all PRs for `accepted_experiences`, `dynamic filtering last batch`, `replacement prompt`, `pending_refs cancel`, `rollout_samples exhausted`, and related combinations. I found no report or fix for this generator-level rollback.

The closest items are not duplicates:

- [#1297](https://github.com/OpenRLHF/OpenRLHF/issues/1297) / [#1300](https://github.com/OpenRLHF/OpenRLHF/pull/1300) fix the synchronous trainer after the generator has already returned a **non-empty** terminal batch. This bug occurs earlier, inside the shared generator, and affects both sync and async paths.
- Open [#1272](https://github.com/OpenRLHF/OpenRLHF/pull/1272) changes the filtering predicate but retains this early-return behavior.
- Open [#1179](https://github.com/OpenRLHF/OpenRLHF/pull/1179) proposes a broader oversampling architecture; it does not provide this targeted fix on current `main`.
- Closed, unmerged [#1190](https://github.com/OpenRLHF/OpenRLHF/pull/1190) changes dispatch architecture but does not fix the terminal accepted-result rollback.
